### ISSUE TITLE
Fix frontend performance issues

### ISSUE THREAD

**Issue by szhu** [Recidiviz/covid19-dashboard#261]
Edit: List of issues in priority order, [as discussed below](https://github.com/Recidiviz/covid19-dashboard/issues/261#issuecomment-622157870):

*   [x] Lazy loading the facilities in the facilities page (#460)
*   [x] Debouncing form inputs for a short duration (~0.3s~ 1s) so that people who type decently fast will get a good experience (#474)
*   [x] Committing form inputs only on enter/unfocus (#474)
*   [ ] Stop using react-spinners
*   [x] Debounce chart and table renders -- probably not needed if inputs are debounced (#474)
*   [ ] Memoize expensive-to-render stuff that has siblings that do constantly need to re-render
*   [x] Consider using an alternative to current GlobalStyles approach (#476)

---

**Is your feature request related to a problem? Please describe.**

A few key interactions on the site are slow, including:

*   Loading the list of facilities is very slow.
*   Loading the list of facilities blocks the UI. The user cannot go to the homepage and immediately load click Add Facility.
*   When editing a facility, typing any text in any of the fields blocks the UI for a bit. It should not block the UI at all.

**Describe the solution you'd like / Describe alternatives you've considered**  
I did some proof of concepts and did some React profiling. These are some solutions that could help. (All need futher investigation/validation.)

Both pages:

*   Consider using an alternative to styled-components GlobalStyles or consider using it a different way. The React profiler suggests that GlobalStyles rendering (at least the way we're currently using it, which is to place it on every page rather than at the root) takes a significant amount of time.

Scenario page:

*   Consider using an alternative to the react-spinners. The React profiler suggests that the emotion css framework that it requires is taking a significant amount of time to render.
*   Consider only rendering the facilities graphs are currently visible in the browser window.
*   Consider adding a short delay (1s?) before fetching any data, so that users who want to load the home page and immediately add a facility will not be blocked.

Facility page:

*   Consider debouncing form field inputs (only commit the input after a short delay).
*   Consider aggressively memoizing components that don't change from render to render (otherwise **every** form field re-renders when editing **any** form field).
*   Consider debouncing the chart and table renders.

I will edit this list as we come up more fixes/if we discover some of these fixes don't result in significant improvement. Please use the comment below to suggest other fixes and other ways to determine what's slow!

---

cc @jessex

**Comment by jessex** [comment]
Thanks a lot for this exploration, @szhu! I've read through this, thought about the value-to-cost ratio, and have some initial thoughts:
* For the scenario page:
  * I am not at all beholden to the react-spinners, though I do think they are very pretty. If there's a lighter-weight alternative that is also pretty, let's switch. But I don't think this is quite as high value as things like shortening the amount of time to load and render the facility rows, or making the facility form (see below) snappier.
  * It feels like a 1s delay is a hack around the fact that we've seemingly coupled components that maybe shouldn't be coupled. If the Add Facility label and button is important enough to load before the facilities load, I think that we should just update the component that holds both and have it return Add Facility _and then a spinner in place of the rows while waiting_ instead of a spinner over the entire thing. That feels like the more canonical React approach to me.
  * I like the idea of only rendering what's currently visible! That would be a nice little win, though I'm not personally familiar with how complex that would be.
  * A question for you, @szhu: based on what you've seen, why do you think that loading the right-side of the page grows as it does when more facility models need to be loaded? Is it the tax of rendering the chart for each row all at once? If so, I'd say only rendering what's currently visible would be the big win here. **To me this might be the highest priority performance issue we can improve.**
* For the facility page:
  * I think the two debouncing items on the facility page will be very helpful. Though it's not the absolute highest priority (I'd say improvements related to loading the scenario page are higher priority), if you are constructing a new facility model from scratch you will likely want to make a variety of edits, and we want that process to feel smooth and enjoyable, not full of friction.
  * I think the memoizing would be a nice follow-up but given the complexity of working with the state management, I wouldn't do it before just **starting with simple debouncing changes and seeing how that works out**.

**Comment by szhu** [comment]
@jessex replies inline!

> I am not at all beholden to the react-spinners, though I do think they are very pretty. If there's a lighter-weight alternative that is also pretty, let's switch. But I don't think this is quite as high value as things like shortening the amount of time to load and render the facility rows, or making the facility form (see below) snappier.

I was thinking about recreating the spinner from scratch, seems like it should take <20 minutes!

> It feels like a 1s delay is a hack around the fact that we've seemingly coupled components that maybe shouldn't be coupled. If the Add Facility label and button is important enough to load before the facilities load, I think that we should just update the component that holds both and have it return Add Facility _and then a spinner in place of the rows while waiting_ instead of a spinner over the entire thing. That feels like the more canonical React approach to me.

I might have not been clear, but when I said "blocking" above, I meant that it's blocking the JS thread, not covering up the button. While the list of facilities loads, there is some computation that's happening that is preventing the user from interacting with any element on the page, even elements that are visible.

(Quick background in just in case you're not familiar with this: JS is single-threaded, so heavy computations will always block like this. It's not a React-specific issue. The general strategy to actually solve this is to split up the computation into smaller chunks with the opportunity for the UI to react in between, which is why I was thinking of just rendering what's visible.)

> I like the idea of only rendering what's currently visible! That would be a nice little win, though I'm not personally familiar with how complex that would be.

Not complex! The approach I've used before is to set a reasonable fixed size for each facility entry (if there isn't one already) and use [react-visiblity-sensor](https://github.com/joshwnj/react-visibility-sensor) to let us know when each facility entry has entered the viewport. Note that our currently set of performance issues doesn't necessitate un-rendering when the facility has left the viewport; we can just leave it rendered.

> based on what you've seen, why do you think that loading the right-side of the page grows as it does when more facility models need to be loaded? Is it the tax of rendering the chart for each row all at once? If so, I'd say only rendering what's currently visible would be the big win here. **To me this might be the highest priority performance issue we can improve.**

Pretty sure it's just the tax of rendering each row. I think it's worth it to play around with the visibility thing to confirm whether it's true.

> I think the two debouncing items on the facility page will be very helpful. Though it's not the absolute highest priority (I'd say improvements related to loading the scenario page are higher priority), if you are constructing a new facility model from scratch you will likely want to make a variety of edits, and we want that process to feel smooth and enjoyable, not full of friction.

👍 

> I think the memoizing would be a nice follow-up but given the complexity of working with the state management, I wouldn't do it before just **starting with simple debouncing changes and seeing how that works out**.

The reason I actually looked into memoizing was because I looked into debouncing and it seemed more complicated than I originally anticipated. I specifically want to make sure that if a user edits a value and clicks Save before the debounce period, then we would save the new value and not the old one -- there's a few ways to ensure this:
- By having a relatively short debounce period (but the shorter the debounce period, the more noticeable the debounce issues). I'd ideally like to have a debounce period of ~0.5s but I think someone might actually go to a facility page to quickly make an edit near the bottom of the page and hit Save immediately after that, possibly within 0.5s. But even 0.5 might be too short -- some users type slower than that.
- By committing the value in the text box only after the user has left it or pressed enter. This is good from an accessibility principle because **it results in a good experience regardless of the user's typing speed**. I kind of want to go with this approach, but the reason I haven't built it out is that I'm thinking through the correct set of props to pass in React to enable this.
- By disabling the Save button during the debounce period.

Memoizing chunks of rendering doesn't have this issue at all, since it doesn't change when the inputs are committed! It was also surprisingly easy to set up (after I figured out how to do it). I tried out using this technique for all of the by-age rows and it did help significantly, but the lag was still noticeable. So I do think it's not enough and we'll still need to debounce.

**Comment by jessex** [comment]
>I might have not been clear, but when I said "blocking" above, I meant that it's blocking the JS thread, not covering up the button. While the list of facilities loads, there is some computation that's happening that is preventing the user from interacting with any element on the page, even elements that are visible.

Yes I understood that, and that JS is single-threaded. However, I wouldn't expect navigation through a link to be blocked (or, at least, certain implementations should be fine) and so by splitting up the lines of separation between components, we should be able to get the component with the link we want rendered before the main thread gets bogged down, thus giving users the ability to navigate away if necessary.

>Not complex! The approach I've used before is... Pretty sure it's just the tax of rendering each row. I think it's worth it to play around with the visibility thing to confirm whether it's true.

Glad to hear it, on both counts! I'd put this at the top of tasks for this issue then.

>By disabling the Save button during the debounce period

I personally would not oppose this! I've seen this behavior in a variety of data-intensive and async-operating apps before, including a variety in the Google suite. 

>Memoizing chunks of rendering doesn't have this issue at all, since it doesn't change when the inputs are committed! It was also surprisingly easy to set up...

Ah! If you have a version of that already working, it would be worth its own PR for review. We just want to make sure we don't end up with stale results on save, but I'm sure you've tested thoroughly.

**Comment by szhu** [comment]
> Yes I understood that, and that JS is single-threaded. However, I wouldn't expect navigation through a link to be blocked (or, at least, certain implementations should be fine) and so by splitting up the lines of separation between components, we should be able to get the component with the link we want rendered before the main thread gets bogged down, thus giving users the ability to navigate away if necessary.

Happy to discuss more, but I disagree with this even theoretically. An extreme example just to clarify: If you have a component that has `while(true);` in its render logic, it doesn't matter where in the render tree this component is placed and how separated the components are -- when the component is being rendered, the entire page will freeze and the user cannot interact with the page at all.

Also thanks for feedback on the rest of the points! Does this sound like a good order of things to try out?
1. Lazy loading the facilities in the facilities page
2. Debouncing form inputs for a short duration (0.3s?) so that people who type decently fast will get a good experience
3. Committing form inputs only on enter/unfocus
4. Stop using react-spinners
5. Debounce chart and table renders
6. Memoize expensive-to-render stuff that has siblings that do constantly need to re-render
7. Consider using an alternative to current GlobalStyles approach

**Comment by jessex** [comment]
That's a perfect order, @szhu!

**Comment by szhu** [comment]
Adding a new item:

> this is super cool @szhu ... there is one question i have beyond the minor issue i have flagged here. One of my suspicions (never fully investigated) was that the major blocker in the facility rows was not the chart rendering itself, but rather the curve function, which is executed synchronously in the `ChartArea` render function when the components are first mounted. I'm just wondering whether that's something you considered here and whether you think it would make an appreciable difference to their performance?
> 
> (I think rendering lots of charts is also probably not super performant regardless of where their data comes from, so I'm not sure which has the greater influence in more extreme cases and for IE users)

https://github.com/Recidiviz/covid19-dashboard/pull/460#pullrequestreview-415508876

**Comment by szhu** [comment]
Closing this issue, since the major items have been fixed! If the two remaining items start to cause significant performance problems in the future, we can re-open.
